core/src/main/java/org/geysermc/geyser/session/PendingMicrosoftAuthentication.java

```diff
@@ -0,0 +1,131 @@
+                .build(new CacheLoader<>() {
+                    @Override
+                    public AuthenticationTask load(@NonNull String userKey) {
+                        return new AuthenticationTask(userKey, timeoutSeconds * 1000L);
+                    }
+                });
+    }
+
+    @SneakyThrows(ExecutionException.class)
+    public AuthenticationTask getOrCreateTask(GeyserSession session) {
+        return authentications.get(
+                Objects.requireNonNull(session.getAuthData(), "authData").xuid()
+        );
+    }
+
+    public static class AuthenticationTask {
+        private final MsaAuthenticationService msaAuthenticationService =
+                new MsaAuthenticationService(GeyserImpl.OAUTH_CLIENT_ID);
+
+        private final String userKey;
+        private final CompletableFuture<MsaAuthenticationService.MsCodeResponse> code;
+        private final CompletableFuture<MsaAuthenticationService> auth;
+        private final long timeoutMs;
+
+        private AuthenticationTask(String userKey, long timeoutMs) {
+            this.userKey = userKey;
+            this.timeoutMs = timeoutMs;
+
+            // Request the code
+            this.code = CompletableFuture.supplyAsync(this::tryGetCode);
+
+            // Once the code is received, continuously try to request the access token, profile, etc
+            this.auth = code.thenApply((code) -> tryLoginContinuously());
+        }
+
+        public CompletableFuture<MsaAuthenticationService.MsCodeResponse> getCode() {
+            return code;
+        }
+
+        public CompletableFuture<MsaAuthenticationService> getAuthentication() {
+            return auth;
+        }
+
+        private MsaAuthenticationService.MsCodeResponse tryGetCode() throws CompletionException {
+            try {
+                return msaAuthenticationService.getAuthCode();
+            } catch (RequestException e) {
+                throw new CompletionException(e);
+            }
+        }
+
+        private MsaAuthenticationService tryLoginContinuously() throws CompletionException {
+            try {
+                long startTime = System.currentTimeMillis();
+                while (true) {
+                    try {
+                        msaAuthenticationService.login();
+                    } catch (AuthPendingException e) {
+                        long deltaTime = System.currentTimeMillis() - startTime;
+                        if(deltaTime > timeoutMs) {
+                            throw new TimeoutException();
+                        }
+                        //noinspection BusyWait
+                        Thread.sleep(1000);
+                        continue;
+                    }
+                    return msaAuthenticationService;
+                }
+            } catch(Exception e) {
+                throw new CompletionException(e);
+            }
+        }
+
+        @Override
+        public String toString() {
```

Comment:
Not entirely sure why we need this method, won't ever be called and when debugging you will be able to see all of the properties anyway.

Reply:
Yeah you're right.

But I left for easier debugging with an IDE: one wouldn't need to open up each object in order to see related `xuid`